added funder links and sample plans to public templates page

briri opened the pull request:
- Added funder links and sample plans to the public DMPTemplates page #880 
- Updated `templates.links` field to be TEXT instead of VARCHAR (was not large enough to store more than one of each link type)
- Added default to `models/template.rb` since a BLOB/TEXT field is not able to have a default value in MySQL

jollopre commented on db/migrate/20171206230107_convert_template_links_to_text.rb:
```diff
@@ -0,0 +1,10 @@
+class ConvertTemplateLinksToText < ActiveRecord::Migration
+  def up
+    # BLOB/TEXT fields cannot have a default value - changing to set default in app/models/template.rb
+    change_column :templates, :links, :text, default: nil
```
I don't see a case right now to change the type from VARCHAR to TEXT. As far as I'm aware both postgresl and mysql can have a variable length VARCHAR which should satisfy the requirements to store the links object as string.

jollopre replied:
Unless we have a string greater than 65535 which probably we never have...

briri replied:
Hmm interesting. If you go in and add 3 funder links and 2 sample plans you get
![screen shot 2017-12-07 at 8 34 50 am](https://user-images.githubusercontent.com/1204467/33726943-60bbe400-db2b-11e7-8200-6e0aff2a0026.png)

All of the varchars in our DBs here are limited to 255 (which was an older limit for MySQL pre 5.0). We're on 5.7 right now which, as you say allows 65k (depending on total row size), so we should be able to increase them on our end.

Its likely something carried over from the original DB we started with at the beginning of the project. I'll have to do some research to determine the best way for us to remove those limitations. In the interim, I'll back out the migration and move of the default to the code.

jollopre replied:
@briri if running your migration solves the problem quickly that's fine but it would be nice to keep VARCHAR since we never going to have huge content for that field.

jollopre commented on app/views/public_pages/template_index.html.erb:
```diff
@@ -18,27 +18,48 @@
+                <br>
                 <%= link_to _('PDF'), template_export_path(template.dmptemplate_id, format: :pdf), target: '_blank' %>
               </td>
+              <td><%= template.org.name %></td>
```
please, can you use the helper I wrote (https://github.com/DMPRoadmap/roadmap/blob/CDL-MVP/app/helpers/template_helper.rb) will reduce verbosity and it has test cases in place.

jollopre commented on app/models/template.rb:
```diff
@@ -170,6 +170,7 @@ def set_creation_defaults
       self.visibility = 1
       self.is_default = false
       self.version = 0 if self.version.nil?
+      self.links = { funder:[], sample_plan:[] } if self.links.nil?
```
the default is set at db level on the migration file, if you don't change the type to TEXT, this line is not needed

briri commented:
removed switch from VARCHAR to TEXT. Going to investigate converting all VARCHARs over from the old pre 5.0 255 limit to 65k. Will document this as other institutions upgrading an existing DMPonline_v4 DB may run into similar issues if they remain on MySQL.

Also updated template_helper's `links_to_elements method` to allow a separator to be passed in

jollopre commented on app/helpers/template_helper.rb:
```diff
@@ -1,8 +1,8 @@
 module TemplateHelper
-  def links_to_a_elements(links)
+  def links_to_a_elements(links, separator = ', ')
```
good one! Even more reusable!

briri commented:
schema file still had TEXT instead of VARCHAR after reverting back. Tests are fixed now
